pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchConverter.java

```diff
@@ -52,12 +52,16 @@ public static boolean isReadableBatch(MessageMetadata metadata) {
         return metadata.hasNumMessagesInBatch() && metadata.getEncryptionKeysCount() == 0;
     }
 
-    public static List<MessageCompactionData> extractMessageCompactionData(RawMessage msg)
+    public static List<MessageCompactionData> extractMessageCompactionData(RawMessage msg, MessageMetadata metadata)
         throws IOException {
         checkArgument(msg.getMessageIdData().getBatchIndex() == -1);
 
         ByteBuf payload = msg.getHeadersAndPayload();
-        MessageMetadata metadata = Commands.parseMessageMetadata(payload);
+        if (metadata == null) {
+            metadata = Commands.parseMessageMetadata(payload);
```

Comment:
One more concern. `Commands.parseMessageMetadata(payload)` will use a shared thread local instance. Is that a problem here?
`Commands.parseMessageMetadata(payload, new MessageMetadata())` would be a way to resolve it if there's a need to have multiple instances in flight.

Reply:
I think it's not a problem.
Metadata is parsed previously when doing compaction in both phase one and phase two.
The null value for metadata only happens in unit tests.